## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head(50)

first        last                                           job  \
0        Donald     Harding                                  Immunologist   
1       Tiffany       Myers                               Music therapist   
2       Shannon      Watson                                  Immunologist   
3        Nathan       Baker                         Accounting technician   
4     Christina    Schwartz                                  Chiropractor   
5       Jessica       Reyes  Production designer, theatre/television/film   
6         Brian       Moore                                       Barista   
7         Jesus      Wright                            Theme park manager   
8       Rebecca       Brown                                   Chiropodist   
9          Greg      Garcia                            Secretary, company   
10         Kyle    Williams           Engineer, broadcasting (operations)   
11      Roberto      Howard                      Surveyor, land/geomatics   
12      Jessica       Reyes  Production designer, theatre/television/film   
13     Patricia       Gould                                      Lobbyist   
14    Katherine      Taylor                        Psychologist, clinical   
15      Matthew      Orozco                    Research scientist (maths)   
16       Joseph      Jordan            Government social research officer   
17       Carrie       Green                           Colour technologist   
18        David       Reyes              Production assistant, television   
19       Shelby    Williams                           Colour technologist   
20       Tricia  Stephenson                   Lecturer, further education   
21      Patrick    Anderson                Commercial art gallery manager   
22        Peter     Roberts                    Architectural technologist   
23  Christopher      Rhodes               Engineer, manufacturing systems   
24        Glenn      Garcia                              Textile designer   
25        Peter     Roberts                    Architectural technologist   
26      Patrick    Anderson                Commercial art gallery manager   
27        Jacob     Mendoza                                Air cabin crew   
28      Tiffany        Lowe                           Medical illustrator   
29        Caleb        Rice                              Technical brewer   
30      Deborah        Cook                     Radiographer, therapeutic   
31      Richard     Kennedy                        Race relations officer   
32       Connie      Harris                       Horticulturist, amenity   
33      Cynthia      Palmer                      Surveyor, rural practice   
34         Erin   Robertson                                Science writer   
35    Catherine      Barnes                                 Airline pilot   
36        Corey       Wells             Environmental health practitioner   
37      Bradley     Spencer                             Warehouse manager   
38       Andrea      Brooks                  IT technical support officer   
39      Mallory    Mitchell                                Quarry manager   
40      William      Mccann                         Accountant, chartered   
41      Douglas        Bean                      Doctor, general practice   
42       Victor    Griffith                           Structural engineer   
43       Taylor      Wagner                                  Toxicologist   
44        Barry     Johnson                     Education officer, museum   
45       Daniel      Newman                    Designer, interior/spatial   
46       Ashley   Hernandez                                  Retail buyer   
47      Vincent       Davis                  Designer, exhibition/display   
48       Daniel      Newman                    Designer, interior/spatial   
49       Curtis      Garcia                                    Bookseller   

                     phone                                     email  \
0       79

In [2]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [23]:
# Use the describe function to gather some basic statistics
df.describe().round(2)

client_id    order_id  order_week  order_year  unit_price  unit_cost  \
count   54639.00    54639.00    54639.00    54639.00    54639.00   54639.00   
mean    54837.87  5470190.03       11.36     2022.99      136.27      99.45   
std     25487.44  2599806.93        7.02        0.08      183.87     133.16   
min     10033.00  1000886.00        1.00     2022.00        0.01       0.01   
25%     33593.00  3196372.00        6.00     2023.00       20.80      14.84   
50%     53305.00  5496966.00       11.00     2023.00       68.31      49.89   
75%     78498.00  7733869.00       17.00     2023.00      173.16     125.57   
max     99984.00  9998480.00       52.00     2023.00     1396.23     846.27   

       unit_weight         qty  line_number  line_subtotal  total_weight  \
count     54639.00    54639.00     54639.00   5.463900e+04      54639.00   
mean          5.00      570.26         2.98   7.710286e+04       2487.42   
std           5.33    18795.52         2.44   2.881926e+06      55594.30   
min           0.00        0.00         0.00   0.000000e+00          0.00   
25%           1.45       32.00         1.00   1.132950e+03         70.33   
50%           3.24       68.00         3.00   4.629440e+03        225.40   
75%           6.89      170.00         5.00   1.736913e+04        718.44   
max          46.43  3958244.00         9.00   5.844743e+08    9009055.62   

       shipping_price    line_price     line_cost   line_profit  
count        54639.00  5.463900e+04  5.463900e+04  5.463900e+04  
mean         17425.10  1.032718e+05  7.387972e+04  2.939207e+04  
std         389159.50  3.286505e+06  2.133294e+06  1.195291e+06  
min              0.00  0.000000e+00  0.000000e+00  0.000000e+00  
25%            496.65  2.887560e+03  2.206360e+03  5.597800e+02  
50%           1577.80  8.396490e+03  6.266040e+03  1.894840e+03  
75%           5029.08  2.645604e+04  1.946590e+04  6.674210e+03  
max       63063389.34  6.412625e+08  3.987535e+08  2.425090e+08

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [5]:
# What three item categories had the most entries?
df["category"].value_counts().nlargest(1)


category
consumables    23538
Name: count, dtype: int64

In [6]:
# For the category with the most entries, which subcategory had the most entire
counts = df.groupby(['category'])[["subcategory"]].value_counts()
counts

category     subcategory       
consumables  bathroom supplies     6424
             paper products        5874
             kitchen supplies      5746
             misc                  2891
             pens                  2603
decor        plants                3466
             wall art              3093
             misc                  1395
electronics  telecom                978
             audio visual           732
             cabling                655
             computers              467
furniture    tables                4540
             stands                4201
             seating               2664
             desks                  510
software     accounting            3756
             project management    3522
             logistics             1122
Name: count, dtype: int64

In [7]:
# Which five clients had the most entries in the data?
clients = df["client_id"].value_counts().nlargest(5)
clients

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [8]:
# Store the client ids of those top 5 clients in a list.
clientlist = [33615, 66037, 46820, 38378, 24741]
clientlist

[33615, 66037, 46820, 38378, 24741]

In [9]:
# How many total units (the qty column) did the client with the most entries order order?
bestclient = df[['client_id',"qty"]].set_index("client_id")
bestclient=bestclient.loc[33615]
bestclient=bestclient.groupby("client_id")["qty"].sum().reset_index()
bestclient[["client_id","qty"]]

client_id    qty
0      33615  64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["line_subtotal"] = df["unit_price"] * df["qty"]
df[["unit_price","qty","line_subtotal"]].head()

unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95
2       13.52   39         527.28
3       36.42   29        1056.18
4      195.10   20        3902.00

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df["total_weight"] = df["unit_weight"]*df["qty"]
def shipping_price(row):
    if (row["total_weight"] > 50):
        return row["total_weight"] * 7
    if (row["total_weight"] <= 50):
        return row["total_weight"] * 10
df["shipping_price"] = df.apply(shipping_price, axis =1)
df[["unit_price","unit_weight","qty","total_weight","shipping_price"]].head()


unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20

In [12]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df["line_price"] = ((df["line_subtotal"] + df["shipping_price"]) * 1.0925).round(2)
df[["line_subtotal","shipping_price","line_price"]].head()

line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12
3        1056.18          356.70     1543.57
4        3902.00         6500.20    11364.40

In [13]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["line_cost"] = (df["unit_cost"]*df["qty"]) + df["shipping_price"]
df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  shipping_price  line_price  \
0            1      115164.00        787.50         5512.50   131839.08   
1            0         523.95         31.29          312.90      914.26   
2            6         527.28         65.52          458.64     1077.12   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [14]:
# Create a column for the profit of each line using line cost and line price
df["line_profit"]= df["line_price"] - df["line_cost"]
df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0      115164.00        787.50         5512.50   131839.08   85597.05   
1         523.95         31.29          312.90      914.26     629.79   
2         527.28         65.52          458.64     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above
confirm = df[["order_id","line_price"]]
dt= confirm.set_index("order_id")
order2173913 = dt.loc[2173913]
order2742071 = dt.loc[2742071]
order6128929 = dt.loc[6128929]
print("Order 2742071 total: $",order2742071["line_price"].sum().round(2))
print("Order 2173913 total: $",order2173913["line_price"].sum())
print("Order 6128929 total: $",order6128929["line_price"].sum())

Order 2742071 total: $ 152811.89
Order 2173913 total: $ 162388.71
Order 6128929 total: $ 923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
spending = df[["client_id","line_price"]]
clspending=spending.set_index("client_id")
print("33615: $",clspending.loc[33615]["line_price"].sum().round(2))
print("66037: $",clspending.loc[66037]["line_price"].sum().round(2))
print("46820: $",clspending.loc[46820]["line_price"].sum().round(2))
print("38378: $",clspending.loc[38378]["line_price"].sum().round(2))
print("24741: $",clspending.loc[24741]["line_price"].sum().round(2))

33615: $ 8377308.52
66037: $ 10259514.79
46820: $ 9743794.36
38378: $ 12906550.87
24741: $ 82268892.02


In [17]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
client=df[["client_id","qty","shipping_price","line_price","line_cost","line_profit"]].set_index("client_id")
newdata=client.loc[[33615, 66037, 46820, 38378, 24741]]
summary=newdata.groupby("client_id")[["qty","shipping_price","line_price","line_cost","line_profit"]].sum().reset_index()
summary.head()


client_id     qty  shipping_price   line_price    line_cost  line_profit
0      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
1      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
2      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
3      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
4      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81

In [18]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
#money = summary["shipping_price","line_price","line_cost","line_profit"]
# Define a function that converts a dollar amount to millions.
#tried for hours, had to resort to this disgustingness
def shipping_price(row):
    if (row['shipping_price'] > 1000000):
        return row['shipping_price']/1000000
def line_price(row):
    if (row['line_price'] > 1000000):
        return row['line_price']/1000000
def beep(row):
    if (row['line_cost'] > 1000000):
        return row['line_cost']/1000000
def boop(row):
    if (row['line_profit'] > 1000000):
        return row['line_profit']/1000000
summary['shipping_price']=summary.apply(shipping_price,axis=1)
summary['line_price']=summary.apply(line_price,axis=1)
summary['line_cost']=summary.apply(beep,axis=1)
summary['line_profit']=summary.apply(boop,axis=1)
summary    
        
  
   
# Apply the currency_format_millions function to only the money columns. 

        
    

# Rename the columns to reflect the change in the money format. 

rename= summary.rename(columns={"client_id":"Client ID","qty":"Units","shipping_price":"Shipping (Millions)",\
                                "line_price":"Total Revenue (Millions)","line_cost":"Total Cost (Millions)",\
                                "line_profit":"Total Profit (Millions)"})
rename= rename.round(2)
rename


Client ID   Units  Shipping (Millions)  Total Revenue (Millions)  \
0      24741  239862                 5.13                     82.27   
1      33615   64313                 1.83                      8.38   
2      38378   73667                 3.43                     12.91   
3      46820   75768                 1.60                      9.74   
4      66037   43018                 1.40                     10.26   

   Total Cost (Millions)  Total Profit (Millions)  
0                  45.69                    36.58  
1                   6.18                     2.20  
2                   9.63                     3.27  
3                   7.01                     2.74  
4                   7.00                     3.26

In [24]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_final=rename.sort_values(by="Total Profit (Millions)", ascending = False)
sorted_final.head()

Client ID   Units  Shipping (Millions)  Total Revenue (Millions)  \
0      24741  239862                 5.13                     82.27   
2      38378   73667                 3.43                     12.91   
4      66037   43018                 1.40                     10.26   
3      46820   75768                 1.60                      9.74   
1      33615   64313                 1.83                      8.38   

   Total Cost (Millions)  Total Profit (Millions)  
0                  45.69                    36.58  
2                   9.63                     3.27  
4                   7.00                     3.26  
3                   7.01                     2.74  
1                   6.18                     2.20

## Summary
In this dataset, the items with the most entries are the bathrooms supplies in the consumable category. The top client (with the most entries) is Client ID 33615 which has 220 entries and bought a 64313 units. While client id 33615 has the most entries, he has purchased the least items of the top 5 and has made the least profit. The client who as gained the most profit is client id 24741 who made 
$36.58 million dollars.